
# Week 01 Assignment  
## Data Quality, Evaluation, Scaling, and Encoding

**Student name: Walid**   

This is a small assignment that connects topics from Module 1, 2, and 3.  
You must complete it in this Colab notebook.

You will need to use concepts that appeared in the videos:
- Module 1 and 2: basic descriptive statistics, proportions, confusion matrix, accuracy, precision, recall
- Module 3: standardization, min max scaling, nominal vs ordinal, one hot encoding, ordinal encoding, Euclidean and Manhattan distance

Please do not use any extra libraries beyond `pandas`, `numpy`.



---
## 0. Setup and Dataset

We will use a dataset that should have columns given below:

- `user_id`  
- `age`  
- `monthly_income` (numeric)  
- `daily_screen_time_min` (numeric)  
- `daily_app_opens` (numeric)  
- `true_label` and `pred_label` for a binary classification task (0 or 1)  
- `satisfaction_level` (for example: `Low`, `Medium`, `High`)  
- `city_type` (for example: `Urban`, `Suburban`, `Rural`)


In [32]:
# Cell 1: Imports
import pandas as pd
import numpy as np

In [33]:
# Cell 2: Load the dataset (Already done for you)
df = pd.read_csv("https://drive.google.com/uc?export=download&id=1OmDDCh4MD1TtvAemnwVDyz5zwCIXJ220")

# Show first few rows
df.head()

,user_id,age,monthly_income,daily_screen_time_min,daily_app_opens,true_label,pred_label,satisfaction_level,city_type
0,1,43,3734.19,109,48,0,0,Medium,Suburban
1,2,49,2594.19,194,7,0,0,Low,Urban
2,3,19,3550.47,146,36,1,0,High,Rural
3,4,19,3821.18,287,14,1,0,High,Suburban
4,5,63,1750.84,66,46,0,0,Medium,Suburban



### 0.1 Check your dataset

1. Confirm that the dataset loaded correctly.  
2. Check that you have at least these columns:  
   - numeric: `age`, `monthly_income`, `daily_screen_time_min`, `daily_app_opens`  
   - labels: `true_label`, `pred_label`  
   - categorical: `satisfaction_level`, `city_type`  



---
## Part A - Module 1 and 2 Review

In this part you will do simple descriptive statistics and basic classification evaluation.



### Q1. Descriptive statistics on a numeric feature

Choose one numeric column, for example `daily_screen_time_min`.


In [34]:
# Q1.1: Choose your numeric column here [We already write this ans]
num_col = "daily_screen_time_min"

df[num_col].describe()

,daily_screen_time_min
count,100.000000
mean,181.890000
std,68.886951
min,60.000000
25%,122.000000
50%,178.000000
75%,243.750000
max,299.000000



> **Q1.2 Short answer: [Marks: 05]**  
> Look at the count, mean, min, max, and standard deviation for your chosen column.  
> In 2 to 3 sentences, comment on what you see.  
> For example, does the max look very far from the mean, or does it look quite close?

Write your answer here:

>  as mean is 181.89 and max is 299.00 , so it seems both have very far range distance. It's because of outliers.
>  
>  



### Q2. Proportion of positive class

Use the `true_label` column, where 1 means "positive" and 0 means "negative".


In [35]:
# Q2.1: Compute proportion of positive class [We already write this ans]
label_col = "true_label"

positive_count = (df[label_col] == 1).sum()
total_count = df.shape[0]
positive_proportion = positive_count / total_count

print("Positive count:", positive_count)
print("Total samples:", total_count)
print("Proportion of positive class:", positive_proportion)

Positive count: 52
Total samples: 100
Proportion of positive class: 0.52



> **Q2.2 Short answer: [5 marks]**  
> In 1 to 2 sentences, explain what this proportion tells you about your dataset.  
> For example, is the dataset balanced between 0 and 1, or is one class much more common?

Write your answer here:

>  According to the table , positive count is 52.As total samples are 100 , so 52% are positive.
>  52% are the majority % of the table.
>  So this table is positive coun t based.




### Q3. Confusion matrix and basic metrics

For this question, use:
- `true_label` as the actual label  
- `pred_label` as the model prediction


In [36]:
# Q3.1: Manually compute TP, TN, FP, FN [We already write this ans]
true_col = "true_label"
pred_col = "pred_label"

tp = ((df[true_col] == 1) & (df[pred_col] == 1)).sum()
tn = ((df[true_col] == 0) & (df[pred_col] == 0)).sum()
fp = ((df[true_col] == 0) & (df[pred_col] == 1)).sum()
fn = ((df[true_col] == 1) & (df[pred_col] == 0)).sum()

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

TP: 28
TN: 27
FP: 21
FN: 24


In [37]:
# Q3.2: Compute accuracy, precision, recall [We already write this ans]
accuracy = (tp + tn) / (tp + tn + fp + fn) if (tp + tn + fp + fn) > 0 else 0

precision = tp / (tp + fp) if (tp + fp) > 0 else 0
recall = tp / (tp + fn) if (tp + fn) > 0 else 0

print("Accuracy:", accuracy)
print("Precision:", precision)
print("Recall:", recall)

Accuracy: 0.55
Precision: 0.5714285714285714
Recall: 0.5384615384615384



> **Q3.3 Short answer: [10 marks]**  
> In 3 to 4 sentences, briefly comment on the model using these three metrics.  
> For example, is the model catching most positives (high recall) or being careful when it predicts positive (high precision)?

Write your answer here:

>  Accuracy shows the overall correctness.
>  precision shows the carefulness of the model predictions of positive cases.
>  recall shows , how actually catching positive cases.



---
## Part B - Module 3: Scaling and Encoding

Now we will pick a few features and apply scaling and encoding.



### Q4. Standardization and Min max scaling

Use one numeric column, `monthly_income`.


In [38]:
# Q4.1: Choose the numeric column [2 marks]
cl = "monthly_income"

In [39]:
# Q4.2: Standardization with z-score [10 marks]
mean_v = df[cl].mean()
std_v = df[cl].std()

df["monthly_income_std"] = (df[cl] - mean_v) / std_v

print("Mean:", mean_v)
print("Standard Deviation:", std_v)
print(df[[cl, "monthly_income_std"]].head())

Mean: 2885.745000000001
Standard Deviation: 898.1246929878644
   monthly_income  monthly_income_std
0         3734.19            0.944685
1         2594.19           -0.324626
2         3550.47            0.740126
3         3821.18            1.041542
4         1750.84           -1.263639


In [40]:
# Q4.3: Min max scaling implementation [10 marks]
min_v = df[cl].min()
max_v = df[cl].max()

df["monthly_income_minmax"] = (df[cl] - min_v) / (max_v - min_v)

print("Minimum:", min_v)
print("Maximum:", max_v)
print(df[[cl, "monthly_income_minmax"]].head())

Minimum: 1000.0
Maximum: 5049.4
   monthly_income  monthly_income_minmax
0         3734.19               0.675209
1         2594.19               0.393685
2         3550.47               0.629839
3         3821.18               0.696691
4         1750.84               0.185420



> **Q4.4 Short answer: [3 marks]**  
> Compare the standardized and min max scaled columns in 2 to 3 sentences.  
> Mention what kind of range each one uses and how the numbers look.

Write your answer here:

>  Standarization values can be between 0 and 1 but both positive and negative values. These are depending on how far these are from mean
>  Min-Max have the fixed range. Between 0 and 1. 0 becomes smallest , 1 becomes largest.
>  



### Q5. One hot and ordinal encoding

We will use:
- `city_type` as a nominal feature  
- `satisfaction_level` as an ordinal feature with order `Low` < `Medium` < `High`  


In [41]:
# Q5.1: One hot encoding for city_type using pandas [10 marks]
one_hot_city = pd.get_dummies(df["city_type"], prefix="city")

print(one_hot_city.head(10))

   city_Rural  city_Suburban  city_Urban
0       False           True       False
1       False          False        True
2        True          False       False
3       False           True       False
4       False           True       False
5        True          False       False
6        True          False       False
7        True          False       False
8       False          False        True
9        True          False       False


In [42]:
# Q5.2: Attach one hot encoded columns to df [5 marks]
df = pd.concat([df, one_hot_city], axis=1)

df.head()

,user_id,age,monthly_income,daily_screen_time_min,daily_app_opens,true_label,pred_label,satisfaction_level,city_type,monthly_income_std,monthly_income_minmax,city_Rural,city_Suburban,city_Urban
0,1,43,3734.19,109,48,0,0,Medium,Suburban,0.944685,0.675209,False,True,False
1,2,49,2594.19,194,7,0,0,Low,Urban,-0.324626,0.393685,False,False,True
2,3,19,3550.47,146,36,1,0,High,Rural,0.740126,0.629839,True,False,False
3,4,19,3821.18,287,14,1,0,High,Suburban,1.041542,0.696691,False,True,False
4,5,63,1750.84,66,46,0,0,Medium,Suburban,-1.263639,0.185420,False,True,False


In [43]:
# Q5.3: Ordinal encoding for satisfaction_level [10 marks]
# Q5.3: Ordinal encoding for satisfaction_level [10 marks]
order = {"Low": 0, "Medium": 1, "High": 2}

df["satisfaction_level_encoded"] = df["satisfaction_level"].map(order)

print(df[["satisfaction_level", "satisfaction_level_encoded"]].head())
df[["satisfaction_level", "satisfaction_level_encoded"]].head()

  satisfaction_level  satisfaction_level_encoded
0             Medium                           1
1                Low                           0
2               High                           2
3               High                           2
4             Medium                           1


,satisfaction_level,satisfaction_level_encoded
0,Medium,1
1,Low,0
2,High,2
3,High,2
4,Medium,1



> **Q5.4 Short answer: [5 marks]**  
> In 2 to 3 sentences, explain why one hot encoding is suitable for `city_type`  
> and why ordinal encoding is suitable for `satisfaction_level`.

Write your answer here:

>  One hot is a nominal features. Which avoid creating false order for city. That's why suitable for city.
>  ordinal encoding is a ordinal features. Which follows a order. That helps to categorized the satisfaction_level by low,high and medium. That's why it's suitable for satisfaction_level
>  



---
## Part C - Module 3: Distances between users

For this small part we will work with vectors based on scaled numeric features.



### Q6. Euclidean and Manhattan distance

Build 2D vectors for user 0 and user 1 using:
- `income_std`  
- `daily_app_opens` (or its min max scaled version if you prefer)


In [44]:
# Q6.1: Build 2D vectors for first two users [We already write this ans]
vec_cols = ["monthly_income", "daily_app_opens"]

v1 = df.loc[0, vec_cols].values
v2 = df.loc[1, vec_cols].values

print("v1:", v1)
print("v2:", v2)

v1: [np.float64(3734.19) np.int64(48)]
v2: [np.float64(2594.19) np.int64(7)]


In [45]:
# Q6.2: Euclidean distance computation [5 marks]
euclidean_distance = np.sqrt(np.sum((v1 - v2) ** 2))

print("Euclidean distance:", euclidean_distance)

Euclidean distance: 1140.7370424422975


In [46]:
# Q6.3: Manhattan distance computation [5 marks]
# Q6.3: Manhattan distance computation [5 marks]
manhattan_distance = np.sum(np.abs(v1 - v2))

print("Manhattan distance:", manhattan_distance)

Manhattan distance: 1181.0



> **Q6.4 Short answer: [5 marks]**  
> Which one is larger in your result, Euclidean or Manhattan distance  
> and why does that usually happen based on their formulas?

Write your answer here:

>   Manhattan distance is larger than Euclidean distance.
>  Because , manhattan adds absolute difference of the distance
>  while euclidean add square root of sum of the squared distance



---
## Final Reflection [10 marks]

> In 4 to 6 sentences, describe how the three modules connect in this assignment.  
> Mention:
> - One idea from Module 1 or 2 that you used  
> - One idea from Module 3 that you used  
> - How these ideas together help you understand a dataset more deeply

Write your reflection here:

>  From module 1 & 2, review model evaluation by using concepts like TP, TN, FP, FN, accuracy, precision and recall.
>  From module 3 , I used standardization, min-max scaling, one-hot encoding and ordinal encoding to prepare the dataset properly
>  Together , these ideas help to understand dataset better by showing how to prepare data, evaluate results and compare users



## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।
